docs/doxygen/bda.md

```diff
@@ -0,0 +1,147 @@
+4. Perform processing steps that only use the data for the current iteration:
+   - Predict
+   - ApplyBeam
+   - ApplyCal
+   - Preflagger
+   - Phaseshift
+   - Filter
+   - Scaledata
+   - Preflagger
+   - UVWFlagger
+5. Perform processing steps that use data from multiple iterations.
+   The steps store data items of the current iteration for
+   use in later iterations. These steps include:
+   - Calibrate
+   - DDECal
+6. The DPPP input should allow specifying the BDA operations above.
+
+### Non-requirements
+
+The following processing steps explicitly do have to support BDA data,
+since it is impossible:
+- Average
+- AOFlagger
+- Interpolate
+- StationAdder
+
+## Design
+
+The current [DPStep](@ref DP3::DPPP::DPStep) class can be reused for supporting
+BDA operations, by extending the process() method with support for BDA data.
+This approach supports all requirements: A step can perform a BDA process()
+call sending BDA data to its next step and implement the BDA process()
+method for receiving BDA data from its previous step.
+
+With this approach, the DPPP input format can remain the same: The steps
+will receive and send data in the requested format.
+
+### DPStep structure
+
+DPSteps can support BDA data using a new process() function for BDA data.
+The default implementatation for this function will throw an exception.
+The existing process() function for regular data, which is now abstract,
+should also get a default implementation that throws an exception. In code:
+
+    class DPStep {
+    public:
+      virtual bool process(DPBuffer) { throw exception; } // Process regular data.
+      virtual bool process(BDABuffer) { throw exception; } // Process BDA data.
+    };
+
+With these two process functions, a DPStep can support
+regular and/or BDA data: Existing DPSteps, which only support regular
+data will remain working. Adding support for BDA data can be done
+incrementally, one DPStep at a time.
+
+A disadvantage of this approach is that DPSteps will become larger and
+more complex: They may get two modes of operation, which
+violates the principle of having a single purpose for each class.
+However, if a DPStep becomes too complex, this design allows
+splitting it into a regular variant and a BDA variant for reducing complexity
+
+### BDA data structure
+
+A crucial point in this design is the BDABuffer which holds BDA data.
+The smallest data element in this structure is a measurement row, which
+contains the measurements for all polarisations in all channels for a single
+baseline and a single time step. Since BDA averages data for multiple channels,
+different rows may have different amounts of channels. In code:
+
+    struct Row {
+      double time;
+      double exposure;
+      Matrix<complex> data(nChannels, nPolarisations);
+      Matrix<bool> flags(nChannels, nPolarisations);
+      Matrix<float> weights(nChannels, nPolarisations);
+      Matrix<bool> fullResFlags(nChannels, nPolarisations);
+    };
```

Comment:
antenna pair should be added

Reply:
Instead, I added an identifier for the baseline.